File: src/Web/src/test/jwtTestUtils.ts
/** UTF-8 JSON → base64url (works in browser + Vitest). */
function toBase64Url(json: string): string {
  const bytes = new TextEncoder().encode(json)
  let binary = ''
  for (const b of bytes) binary += String.fromCharCode(b)
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

/** Matches backend migration default tenant for local/dev JWT stubs. */
const DEFAULT_TEST_TENANT_ID = 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa'

/** Minimal unsigned JWT for tests (payload only verified client-side). */
export function makeUnsignedJwt(payload: Record<string, unknown>): string {
  const merged = { tenant_id: DEFAULT_TEST_TENANT_ID, ...payload }
  const header = toBase64Url(JSON.stringify({ alg: 'none', typ: 'JWT' }))
  const body = toBase64Url(JSON.stringify(merged))
  return `${header}.${body}.sig`
}

export function expSoon(secondsFromNow = 3600): number {
  return Math.floor(Date.now() / 1000) + secondsFromNow
}

File: src/Tests/Web.Blazor/TestHelpers/JwtTestUtils.cs
using System.Text;
using System.Text.Json;

namespace SafetyScale.Tests.Web.Blazor.TestHelpers;

/// <summary>Port of React <c>jwtTestUtils.ts</c> for Blazor JWT/session tests.</summary>
public static class JwtTestUtils
{
    public const string DefaultTestTenantId = "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa";

    public static string MakeUnsignedJwt(IReadOnlyDictionary<string, object?> payload)
    {
        var merged = new Dictionary<string, object?>(payload);
        merged.TryAdd("tenant_id", DefaultTestTenantId);

        var header = ToBase64Url("""{"alg":"none","typ":"JWT"}""");
        var body = ToBase64Url(JsonSerializer.Serialize(merged));
        return $"{header}.{body}.sig";
    }

    public static long ExpSoon(int secondsFromNow = 3600) =>
        DateTimeOffset.UtcNow.ToUnixTimeSeconds() + secondsFromNow;

    private static string ToBase64Url(string json)
    {
        var bytes = Encoding.UTF8.GetBytes(json);
        return Convert.ToBase64String(bytes)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }
}
